packages/orderbook/src/order_provider/sra_websocket_order_provider.ts

```diff
@@ -0,0 +1,182 @@
+        const currentOrders = this._orderStore.getOrderSetForAssetPair(assetPairKey);
+        const newOrders = new OrderSet(orders);
+        const diff = currentOrders.diff(newOrders);
+        this._updateStore({
+            added: diff.added,
+            removed: diff.removed,
+            assetPairKey,
+        });
+    }
+
+    private async _syncOrdersInOrderStoreAsync(): Promise<void> {
+        for (const assetPairKey of this._orderStore.keys()) {
+            const [assetDataA, assetDataB] = OrderStore.assetPairKeyToAssets(assetPairKey);
+            await this._fetchAndCreateSubscriptionAsync(assetDataA, assetDataB);
+        }
+    }
+
+    /**
+     * Creates a new websocket orders channel.
+     */
+    private async _createOrdersChannelAsync(): Promise<OrdersChannel> {
+        const ordersChannelHandler: OrdersChannelHandler = {
+            onUpdate: async (_channel, _opts, apiOrders) => this._handleOrderUpdates(apiOrders),
+            // tslint:disable-next-line:no-empty
+            onError: (_channel, _err) => {},
+            onClose: async () => {
+                // Do not reconnect if destroyed
+                if (this._isDestroyed) {
+                    return;
+                }
+                // Re-sync and create subscriptions
+                await utils.attemptAsync<boolean>(async () => {
+                    this._ordersChannel = undefined;
+                    await this._syncOrdersInOrderStoreAsync();
+                    return true;
+                });
+            },
+        };
+        try {
+            return await ordersChannelFactory.createWebSocketOrdersChannelAsync(
+                this._websocketEndpoint,
+                ordersChannelHandler,
+            );
+        } catch (e) {
+            // Provide a more informative error
+            throw new Error(`Creating websocket connection to ${this._websocketEndpoint}`);
+        }
+    }
+
+    /**
+     * Handles updates from the websocket, adding new orders and removing orders
+     * which have remainingFillableTakerAssetAmount as 0.
+     * @param orders the set of API Orders returned from the websocket channel
+     */
+    private _handleOrderUpdates(orders: APIOrder[]): void {
+        const addedRemovedByKey: { [assetPairKey: string]: AddedRemovedOrders } = {};
+        for (const order of orders) {
+            const assetPairKey = OrderStore.getKeyForAssetPair(order.order.makerAssetData, order.order.takerAssetData);
+            if (!addedRemovedByKey[assetPairKey]) {
+                addedRemovedByKey[assetPairKey] = { added: [], removed: [], assetPairKey };
+            }
+            const addedRemoved = addedRemovedByKey[assetPairKey];
+            // If we have the metadata informing us that the order cannot be filled for any amount we don't add it
+            const remainingFillableTakerAssetAmount = (order.metaData as any).remainingFillableTakerAssetAmount;
```

Comment:
When would we not have the metadata?

Reply:
We do not define what is inside of the `metaData` as per the SRA spec. Radar and Launch kit return `remainingFillableTakerAssetAmount` but it is not enforced by the spec.